packages/desktop-gui/cypress/integration/settings_spec.js

```diff
@@ -181,6 +182,79 @@ describe('Settings', () => {
           expect(this.ipc.externalOpen).to.be.calledWith('https://on.cypress.io/guides/configuration')
         })
       })
+
+      it('displays null when no env settings are found', function () {
+        const setConfigEnv = (config, v) => {
+          return flow([
+            merge(config),
+            set('resolved.env', v),
+          ])({})
+        }
+
+        this.ipc.openProject.onCall(1).resolves(setConfigEnv(this.config, null))
+        this.ipc.onConfigChanged.yield()
+
+        cy.contains('.line', 'env:null')
+
+        this.ipc.openProject.onCall(2).resolves(setConfigEnv(this.config, {}))
+        this.ipc.onConfigChanged.yield()
+
+        cy.contains('.line', 'env:null')
+
+        this.ipc.openProject.onCall(3).resolves(setConfigEnv(this.config, undefined))
+        this.ipc.onConfigChanged.yield()
+
+        cy.contains('.line', 'env:null')
```

Comment:
i don't believe you're coordinating the stubs + the cypress commands correctly here because the cypress commands are async and the `this.ipc.onConfigChanged.yield()` is synchronous. I believe it's synchronously firing the `onConfigChanged` immediately (three times), and so the final `undefined` value is being flushed to the DOM, and then the three `cy.contains(...)` are running only against the last render of the DOM (which is `undefined`).

To fix this you need to use a `.then()` so that the IPC events are fired after the `cy.contains()`

Reply:
Thanks for pointing this out and going over it with me. Consequently, the implementation passes for each use case, `null`, `undefined`, and `{}`. However, even with the new code changes to the test that we worked on, I get strange behavior. I broke them into 3 separate E2E test cases, but I don't much like it. I'd love if anyone has any insights on how to combine them into a single test case.